### ISSUE TITLE
Metabolomics improvements 

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#5859]
Evaluate
SLAW: A Scalable and Self-Optimizing Processing Workflow for Untargeted LC-MS
https://pubs.acs.org/doi/10.1021/acs.analchem.1c02687
and check if our novel workflow closes some of the gaps they mentioned.

**Comment by jpfeuffer** [comment]
Specifically, we are missing:
- Quality control (e.g. looking and browsing through extracted isotope traces in each found feature)
- Parameter optimization (the noise threshold makes such a HUGE difference, that there cannot be a default) https://github.com/OpenMS/OpenMS/discussions/5388
- Adduct decharging takes too long. Can we improve our default parameters or speed it up.
- We seem to be faster than other methods (which is surprising, I really have troubles when lowering the noise threshold) see also the discussion I opened
- Is our alignment and linking memory efficient? (see #5756)

see also #5789 #5336 #4069 #3878

**Comment by timosachsenberg** [comment]
Regarding "Adduct decharging takes too long. Can we improve our default parameters or speed it up." @eeko-kon raised some doubts. Didn't we reduce the huge set of adducts at some point?

**Comment by eeko-kon** [comment]
FeatureFinder and MetaboAdduct decharger are really fast, at least with my samples (which are quite complex). I have now run their workflow (actually using docker is a super good idea for the snakemake-OpenMS workflow) , it was very easy to set up. I just got the tables and will investigate them for the feature information and then also compare the speed. I m only using 3 files for now for comparison.

I think the QC idea is really nice

Parameter optimization really depends on the data (e.g. QTOF noise threshold would be way lower than an Orbitrap instrument) so really people cannot avoid this step - unless OpenMS generates the parameters after investigating the intrument and ionisation techniques aka noise threshold, adduct information, etc

**Comment by eeko-kon** [comment]
Ok here are the stats so far for 3 identical files, using 3 cores in my Mac:

OpenMS (snakemake)
1. FeatureFinderMetabo -in {input} -out {output} -algorithm:common:noise_threshold_int "1.0e04" -algorithm:mtd:mass_error_ppm "10.0" -algorithm:epd:width_filtering "fixed" -algorithm:ffm:isotope_filtering_model "none" -algorithm:ffm:remove_single_traces "true" -algorithm:ffm:report_convex_hulls "true" — 

FeatureFinderMetabo took 1.63 s (wall), 1.54 s (CPU), 0.09 s (system), 1.45 s (user); Peak Memory Usage: 60 MB.
FeatureFinderMetabo took 1.85 s (wall), 1.78 s (CPU), 0.11 s (system), 1.67 s (user); Peak Memory Usage: 65 MB.
FeatureFinderMetabo took 1.68 s (wall), 1.60 s (CPU), 0.09 s (system), 1.51 s (user); Peak Memory Usage: 66 MB.
Total: 5.16 seconds

2. MetaboliteFeatureDeconv 
MetaboliteAdductDecharger took 0.51 s (wall), 0.49 s (CPU), 0.06 s (system), 0.43 s (user); Peak Memory Usage: 30 MB.
MetaboliteAdductDecharger took 0.37 s (wall), 0.28 s (CPU), 0.05 s (system), 0.23 s (user); Peak Memory Usage: 25 MB.
MetaboliteAdductDecharger took 0.37 s (wall), 0.31 s (CPU), 0.05 s (system), 0.26 s (user); Peak Memory Usage: 25 MB.
Total: 1.25 seconds

3. TextExporter (csv table) 

TextExporter took 0.29 s (wall), 0.29 s (CPU), 0.01 s (system), 0.28 s (user); Peak Memory Usage: 18 MB.
TextExporter took 0.20 s (wall), 0.20 s (CPU), 0.00 s (system), 0.20 s (user); Peak Memory Usage: 17 MB.
TextExporter took 0.19 s (wall), 0.19 s (CPU), 0.01 s (system), 0.18 s (user); Peak Memory Usage: 17 MB.
Total 0.68 seconds

4. MapAlignerPoseClustering took 1.13 s (wall), 1.10 s (CPU), 0.11 s (system), 0.99 s (user); Peak Memory Usage: 26 MB.

SLAW (Docker+Nextflow)

1. FeatureFinderMetabo -in "/input/20220204_NBC_Plate26_POS_00906_DNPM_HLB100.mzML" -out "/output/OPENMS/peaktables/peaktable_2_62db2b41b9c67b55fde0223197099ea0.featureXML" -algorithm:common:chrom_peak_snr 3.0 -algorithm:epd:masstrace_snr_filtering true -algorithm:epd:max_fwhm 6.0  -algorithm:epd:min_fwhm 1.2 -algorithm:common:chrom_fwhm 2.16 -algorithm:ffm:use_smoothed_intensities true -algorithm:epd:width_filtering fixed -algorithm:mtd:mass_error_ppm 15.0 -algorithm:mtd:quant_method area -algorithm:mtd:trace_termination_criterion outlier -algorithm:mtd:trace_termination_outliers 5 -algorithm:mtd:min_trace_length 1.2

Warning: ElutionPeakDetection received the unknown parameter 'noise_threshold_int'! --there is an error here so this param is not passed actually

FeatureFinderMetabo took 6.57 s (wall), 6.28 s (CPU), 0.13 s (system), 6.15 s (user).
FeatureFinderMetabo took 7.56 s (wall), 7.30 s (CPU), 0.12 s (system), 7.18 s (user).
FeatureFinderMetabo took 8.56 s (wall), 8.25 s (CPU), 0.21 s (system), 8.04 s (user).
I guess this is due to noise threshold zero.

2. Alignment
STEP: alignment TOTAL_TIME:61.56s LAST_STEP:14.43s

3. Gap filling and isotopic pattern extraction finished. (This is like a missing value step)
STEP: gap-filling TOTAL_TIME:61.57s LAST_STEP:0.01s

4. Annotation finished (adducts)
STEP: annotation TOTAL_TIME:61.61s LAST_STEP:0.03s

**Comment by jpfeuffer** [comment]
> FeatureFinder and MetaboAdduct decharger are really fast, at least with my samples (which are quite complex). I have now run their workflow (actually using docker is a super good idea for the snakemake-OpenMS workflow) , it was very easy to set up. I just got the tables and will investigate them for the feature information and then also compare the speed. I m only using 3 files for now for comparison.

see my last comment

> I think the QC idea is really nice

We could discuss at some point which QC metrics are interesting for Metabo and how to create them

> Parameter optimization really depends on the data (e.g. QTOF noise threshold would be way lower than an Orbitrap instrument) so really people cannot avoid this step - unless OpenMS generates the parameters after investigating the intrument and ionisation techniques aka noise threshold, adduct information, etc

True! And it is indeed what I was thinking of. If FFM is fast, we could run it multiple times to improve on a certain objective in QC data or something like this. Or estimate some parameters from the data. (Instrument should also be in the mzML file)

**Comment by eeko-kon** [comment]
The requantification step for OpenMS:
FeatureFinderMetaboIdent took 6.44 s (wall), 5.57 s (CPU), 0.26 s (system), 5.31 s (user); Peak Memory Usage: 118 MB.

I guess these are different workflows in terms of gap-filling which probably takes place in the aligned featureXML files.

**Comment by jpfeuffer** [comment]
> The requantification step for OpenMS: FeatureFinderMetaboIdent took 6.44 s (wall), 5.57 s (CPU), 0.26 s (system), 5.31 s (user); Peak Memory Usage: 118 MB.
> 
> I guess these are different workflows in terms of gap-filling which probably takes place in the aligned featureXML files.

Aah you are doing requantification already? That's cool.

**Comment by eeko-kon** [comment]
Hm all my files are max 40MB each. I m trying to download the ones you sent - any idea about their size?

**Comment by jpfeuffer** [comment]
> Hm all my files are max 40MB each. I m trying to download the ones you sent - any idea about their size?

Unzipped as mzml they are around 160MB: https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Tutorials/Data/MetaboMS1Benchmark/

**Comment by eeko-kon** [comment]
Ok let me try with those.

**Comment by eeko-kon** [comment]
For the SA1 file (170MB)
FeatureFinderMetabo took 38.19 s (wall), 37.69 s (CPU), 2.61 s (system), 35.08 s (user); Peak Memory Usage: 1362 MB.

**Comment by jpfeuffer** [comment]
😆  what? which params? e.g. noise_threshold, snr, fwhm?

**Comment by eeko-kon** [comment]
The same ones I used above. I also tried with noise thr 1.0 and...:

FeatureFinderMetabo took 01:13 m (wall), 01:12 m (CPU), 3.62 s (system), 01:08 m (user); Peak Memory Usage: 1650 MB.

what params are you using?

**Comment by eeko-kon** [comment]
Params:

`FeatureFinderMetabo -in {input} -out {output} -algorithm:common:noise_threshold_int "1.0e04" -algorithm:mtd:mass_error_ppm "10.0" -algorithm:epd:width_filtering "fixed" -algorithm:ffm:isotope_filtering_model "none" -algorithm:ffm:remove_single_traces "true" -algorithm:ffm:report_convex_hulls "true"`

switched to -algorithm:common:noise_threshold_int "1.0" to check the time it needs

**Comment by jpfeuffer** [comment]
Right now, I have noise 80000 just to make my workflow run through :D 
But I also have some very loose RT settings, since I saw quite some small traces in the data.

![Bildschirmfoto 2022-02-25 um 16 40 32](https://user-images.githubusercontent.com/8102638/155743763-21ada851-b522-4b85-928f-e73de32cb10d.png)

**Comment by eeko-kon** [comment]
`-algorithm:ffm:remove_single_traces "true"`
and
`-mz_scoring_by_elements "true"`
are the differences that I can see in our parameters. Changing those, again, my run finished quite fast:

FeatureFinderMetabo took 42.84 s (wall), 41.93 s (CPU), 3.26 s (system), 38.67 s (user); Peak Memory Usage: 1398 MB.

**Comment by jpfeuffer** [comment]
Did you change the min_trace_length and min_fwhm manually? I couldn't find those parameters in the snakemake files. They are medium important for this dataset. There are some ground truth features in the csv file which should be found.

**Comment by eeko-kon** [comment]
No, those parameters are only used in SLAW. I haven't used them in my workflow at all.

**Comment by eeko-kon** [comment]
So using SLAW, I couldn't find the compounds of interest or as you say the ground truth features for any of the three files. I think the parameters they are using are simply not working with my data. I will look a bit more into it tomorrow.

**Comment by jpfeuffer** [comment]
> No, those parameters are only used in SLAW. I haven't used them in my workflow at all.

Ah, my guess is that those parameters are responsible for higher runtimes on my data though. Can you try running with those parameters in your pipeline (or just on one file with just FFMetabo) when you find time?

**Comment by eeko-kon** [comment]
I just did- these are all the parameters they are using :

`resources/OpenMS-2.7.0/bin/FeatureFinderMetabo -in {input} -out {output} -algorithm:common:noise_threshold_int "0" -algorithm:mtd:mass_error_ppm "15.0" -algorithm:epd:width_filtering "fixed" -algorithm:common:chrom_peak_snr 3.0 -algorithm:epd:masstrace_snr_filtering -algorithm:epd:max_fwhm 6.0 -algorithm:epd:min_fwhm 1.2 -algorithm:common:chrom_fwhm 2.16 -algorithm:ffm:use_smoothed_intensities true -algorithm:epd:width_filtering fixed -algorithm:mtd:mass_error_ppm 15.0 -algorithm:mtd:quant_method area -algorithm:mtd:trace_termination_criterion outlier -algorithm:mtd:trace_termination_outliers 5 -algorithm:mtd:min_trace_length 1.2 -threads {threads}`
[parameters.txt](https://github.com/OpenMS/OpenMS/files/8146089/parameters.txt)

FeatureFinderMetabo took 47.68 s (wall), 01:05 m (CPU), 2.94 s (system), 01:02 m (user); Peak Memory Usage: 1164 MB.

However, I m getting none of the ground truth features this way.

On another note:

In the file parameters.txt, they introduce all the parameters, which is actually really nice (e.g. under peak picking):
```
noise_level_ms1:
description: Integer. An absolute noise threshold to be used on MS1. Any points in a RAW file below this file will be discarded. This should not be too high, as low intensity part can still occurs at the border of a peak. priority: ESSENTIAL value: 10000 - **this info is def not passed since I have a table of features with lower intensities than 10^4**

noise_level_ms2:
description: Integer. An absolute noise threshold to be used on MS2. Any points in a RAW file below this file will be discarded.
priority: ESSENTIAL
value: 0

peaks_deconvolution:
SN:
description: Float. (All peakpicker) The signal-over-noise threshold.
priority: HIGH
range:
- -5
- 15
value: 3

ms2_mz_tol:
description: Float. (All peakpickers) The tolerance in mass in Dalton for MS-MS pairing.
priority: LOW
value: 0.1

ms2_rt_tol:
description: Float. (All peakpickers) The tolerance in rt for MS-MS pairing in minutes.
priority: LOW
value: 0.1

noise_level:
description: Float. (All peakpicker) The mininimum feature height for a deconvoluted peak to be kept.
priority: HIGH
value: 0

peak_width:
description: Floats. (All peakpicker) The minimum and maximum width of a peak in minutes. This parameter is highly dependent of you chromatography setup and should be the first target of optimization.
priority: ESSENTIAL
range:
max:
- 0.05
- 0.5
min:
- 0.015
- 0.15
value:
- 0.02
- 0.1

peak_width_fac:
description: Float. (OPENMS only) FLoat between 0 and 1 a factor indicating the main peakwidth for feature finder metabo as peak_width[0]+peak_width_fac*(peak_width[1]-peak_width[0])
priority: HIGH
range:
- 0.0
- 1.0
value: 0.2

peaktable_filter:
description: String. A string describing the postprocessing to apply to the peaktable. See the wiki for examples. The default is to only take the 30000 most intense peaks.
priority: LOW
type: str
value: absolute_intensity top 30000
```
etc..

**Comment by jpfeuffer** [comment]
Ok, then it must be something on my side. I will investigate here. I usually find at least 80% of the 1100 ground truth features but I don't remember the exact way of how I am matching them.

**Comment by eeko-kon** [comment]
Ah, got it nevermind :)

**Comment by jpfeuffer** [comment]
Oh yes I should add the Metabolites CSV to our HTTP.

**Comment by eeko-kon** [comment]
(the naming is a bit misleading - they are both OpenMS , just the FFM has different params - by OPENMSwf I mean the ones I currently use and by SLAW the ones SLAW uses)

**Comment by jpfeuffer** [comment]
I just rebuilt FFM in Release mode and ran `./bin/FeatureFinderMetabo -ini ~/Downloads/FFM_1_test_KNIMEWF_sensitivity.ini -in SA1.mzML -out test.featureXML`. Taking almost an hour already.

[FFM_1_test_KNIMEWF_sensitivity.ini.zip](https://github.com/OpenMS/OpenMS/files/8148984/FFM_1_test_KNIMEWF_sensitivity.ini.zip)

I can usually identify around 600 ground truth features correctly..
https://mab.to/KhmnZRt0M

**Comment by eeko-kon** [comment]
So every file has all 600 compounds? That's what I don't understand. Or is it undefined in the experiment? Either way, I have a total of 487 annotations using the SLAW parameters and 47 using the ones from my workflow. I m just not sure how to tell that this is correct.

**Comment by eeko-kon** [comment]
Also, can it be perhaps because you are using KNIME with such large files that it is that slow? Otherwise I don't understand why..

**Comment by jpfeuffer** [comment]
Yes so according to the publication 1100 compounds were spiked in in each sample. 1,2,3 are replicates. The concentrations differ between A and B for some of the compounds. Around 800 were validated manually by the authors, so they should be findable in each sample 

My command above was run from the command line, not KNIME. Does your workflow maybe do a second peakpicking? What if you run my command on the command line outside of the workflow?

**Comment by eeko-kon** [comment]
Ok that's a lot. Thanks for that, was a bit tired yesterday and did not read thoroughly. Got it. You are simply using the ini file from KNIME. 

Alright, so for some reason, using the params directly in the CLI is much faster:
`FeatureFinderMetabo took 44.02 s (wall), 43.70 s (CPU), 1.33 s (system), 42.37 s (user); Peak Memory Usage: 1129 MB.
`
Whereas with your command- introducing the params through the ini file takes forever. I m also getting this warning:

`Warning: Parameters file version (2.6.0-pre-develop-2021-06-08) does not match the version of this tool (2.7.0).
`

I think that causes the issue because I created a .ini file from scratch in TOPPView and it took exactly the same time as introducing them in the CLI directly
[Params_SLAW.ini.zip](https://github.com/OpenMS/OpenMS/files/8151512/Params_SLAW.ini.zip)
.

**Comment by eeko-kon** [comment]
(I will send you another report with the results I am getting through my workflow vs SLAW)

**Comment by jpfeuffer** [comment]
@timosachsenberg or someone else, can you run the data quickly through FFMetabo with my ini and see if it takes 40 seconds or 4 hours?

**Comment by eeko-kon** [comment]
@jpfeuffer Ok so I think we misunderstood eachother here: I opened your ini file in TOPPView and I noticed that you have completely different parameters than the ones used in SLAW (or of course the ones I m using). These are your parameters:

`./bin/FeatureFinderMetabo -in SA1.mzML  -out SA1.featureXML -algorithm:common:noise_threshold_int "1.0e04" -algorithm:mtd:mass_error_ppm "10.0" -algorithm:epd:width_filtering "fixed" -algorithm:ffm:isotope_filtering_model "none" -algorithm:ffm:remove_single_traces "false" -algorithm:ffm:report_convex_hulls "true" -algorithm:epd:max_fwhm 40.0 -algorithm:epd:min_fwhm 1.0 -algorithm:mtd:min_trace_length 0.5 -algorithm:mtd:max_trace_length 80.0 -algorithm:ffm:isotope_filtering_model "none"`

And these are the ones used in SLAW:

`./bin/FeatureFinderMetabo -in SA1.mzML  -out SA1.featureXML -algorithm:common:noise_threshold_int "10.0" -algorithm:mtd:mass_error_ppm "15.0" -algorithm:epd:width_filtering "fixed" -algorithm:common:chrom_peak_snr 3.0 -algorithm:epd:masstrace_snr_filtering -algorithm:epd:max_fwhm 6.0 -algorithm:epd:min_fwhm 1.2 -algorithm:common:chrom_fwhm 2.16 -algorithm:ffm:use_smoothed_intensities true -algorithm:epd:width_filtering fixed -algorithm:mtd:mass_error_ppm 15.0 -algorithm:mtd:quant_method area -algorithm:mtd:trace_termination_criterion outlier -algorithm:mtd:trace_termination_outliers 5 -algorithm:mtd:min_trace_length 1.2`

I think the main differences are that you have a very short min_trace_length, a long max_trace_length and also do not remove single mass traces (all this from mass trace detection which takes very long) - all this probably picks up noise? or most likely contaminants that elute throughout a long RT? (correct me if I misinterpret the parameters) but somehow it works very well for this dataset. I'm letting it run overnight to verify the results.

I thought that you were using the parameters used in SLAW -that was the misunderstanding :D 

For now what I can say with facts is that:
- Using my workflow, I could annotate 768 compounds (Thermo QE files)
- Using my workflow with the SLAW params (FFM), I could annotate 753 compounds (Thermo QE files)
- Using the SLAW workflow, I could identify zero features (empty peak files), except from file SB5.mzML (30,000 features) where I could annotate 686 compounds  (Thermo QE files)

With this dataset, my workflow was the fastest and it got the highest number of annotations.
I think this is actually a great dataset to use as proof-of-concept in the publication.

**Comment by jpfeuffer** [comment]
yes 768 seems pretty good. In the manuscript, other tools got between 748 and 820 of the features.

**Comment by eeko-kon** [comment]
Then perhaps if we tweak the params and requantify we could reach a higher number of hits.
